Release 2.2.0

RlanderRISCSW opened the pull request:
Pull request for changes on existing TIGL code from Airbus Defence and Space.

Major changes:
- Wing Profile code changed for requesting profile geometry with sharp and flat trailing edge
- New class CCPACSModel for use as root component
- Support for "simple" transformations in CTiglTransformation for performance increase
- extended implementation of CCPACSWingComponentSegment

rainman110 commented:
Thanks for this pull request. Just a general comment to the commit. This is a pretty huge commit, making it really hard to review. It would be great, if you would split it up into smaller commits in future.

By scanning quickly over the code, I saw a range of potential problems, which we should discuss:

- [ ] Renaming of ```CCPACSComponentSegemt::findSegment``` into ```CCPACSComponentSegemt::FindSegment```. It would be good, if existing signatures are not changed without good reason.

- [ ] Duplicate IGES export code (CTiglExportIGES (DLR) vs. CTiglIGESWriter (Airbus)). It would be good if you could integrate your IGES stuff into CTiglExportIGES. BTW, what does  ```XSAlgo::AlgoContainer()->ProcessShape``` do?

- [ ] Duplicate common functions: src/common/CTiglCommon vs src/common/tiglcommonfunctions. 

- [ ] Why ```getProfiles()``` methods in CCPACSFuselages and CCPACSWings? This is somewhat duplicate to getProfile(index) and blows up the class interface.

- [ ] Why the destiction between open and closed wing profiles? The CPACS file should specify, whether a wing profile is open or closed (last and first point are identical).

- [ ] The class CCPACSWingComponentSegment seems to be pretty blown up with methods. Many of them seem to be convenience methods, which is not considered as good design. Is this really necessary. Wouldn't a helper class not serve the same purpose?

Also, it would be great, if pull request would compile successfully. This would allow us to see also, whether all test cases still run successfully. Currently, travis fails to compile: https://travis-ci.org/DLR-SC/tigl/builds/124866357

You can also enable travis CI in you github fork, such that you directly get feedback after commiting.

RlanderRISCSW commented:
Sorry for this big chunk, i'll try to split up the changes in smaller pieces in future.
Comments on your points:

- Renaming of `CCPACSWingComponentSegemt::findSegment `into `CCPACSWingComponentSegemt::FindSegment`: renaming was probably done out of consistency reasons (public methods seem to be all upper case), but is of course not necessary. I will revert this change in order to not break the API
- Duplicate IGES export code (CTiglExportIGES (DLR) vs. CTiglIGESWriter (Airbus)): CTiglIGESWriter is used by additional methods in CTiglExportIges, which will be merged in a separate step. So i'll remove CTiglIGESWriter for now
- Duplicate common functions: CTiglCommon was created at a time where tiglcommonfunctions was not existing. I'll merge the functionality, which is currently used, into tiglcommonfunctions
- Why getProfiles() methods in CCPACSFuselages and CCPACSWings: external functionality requires adding/removal of profiles during runtime. This was realized by adding new methods `AddProfile`and `DeleteProfile` to the `CCPACSWingProfiles` and `CCPACSFuselageProfiles` classes. In order to get access to these classes the getters were implemented in `CCPACSWings` and `CCPACSFuselages` classes. Proposals for alternative implementations are welcome.
- open and closed wing profiles are required for different meshes. The wing geometry (shape) is used as a basis for meshing for the generation of analysis meshes for FEM, but also for aero codes. FE meshes require a sharp trailing edge, while aero meshes require an edge at the trailing edge of the wing. The same CPACS model should be used for the generation of both meshes, so the code for opening the closed profile was added. The current modification we did in the profile code is not CPACS conform, since it actively changes the profile points (e.g. extends or trims to x==1). My proposal would be that the modification of profile points (if required) is done by us outside of the TIGL code, so we can ensure in our application that all profiles have a sharp trailing edge (required for FE mesh) without modifying the existing profile code in TIGL. But we need the logic for getting a modified profile geometry with a "flat" trailing edge for the generation of the aero mesh. 
- The class CCPACSWingComponentSegment seems to be pretty blown up with methods: correct, sepearation in helper classes will be investigated

The current version of the source code compiled perfectly locally on my machine. Probably the used OCE version 0.16.1 is too old for the release-2.2.0 branch, which version is required? I have no experience with travis CI, but will have a look at it

rainman110 commented:
> Probably the used OCE version 0.16.1 is too old for the release-2.2.0 branch, which version is required? 

Minimum required version is OCE 0.15 / OCCT 6.7.0.

The code should also compile up to 6.9.1. The TiGL master branch works also on OCCT 7.0.0.

Travis uses OCE 0.17 / OCCT 6.8.0 packages from here: https://build.opensuse.org/package/show/home:martinsiggel/OCE

To enable travis for TiGL, just create a travis account and enable it for TiGL. It will then automatically build for each commit. You don't need to configure anything, as this is already done in .travis.yml.

rainman110 commented:
How are we dealing with the failing tests? Are you working on it? Do you need help?

RlanderRISCSW commented:
I'm currently trying to run the tests on my machine in order to find out the reason for the failed tests. Thx, I'll contact you as soon as i need help.

RlanderRISCSW commented:
Last commit is a proposal for fixing the test case tiglExportSimple.check_face_traits: The problem arises because we introduced the new class CCPACSModel as root component in the CTiglAbstractPhysicalComponent tree. This was done in order to support the definition of CPACS models, where absolute coordinates were used for the wing or fuselage transformations. Using the CCPACSModel as root component allows to keep the existing tree structure logic. The test failed because the CCPACSModel does not have a shape itself, so the generation of the fused plane geometry (CTiglFusePlane) did not support components without shape.
The proposed fix allows now for components without shape.

rainman110 commented:
This fix looks okay. I somehow don't really like to have an separate CCPACSModel class, as this should be the CCPACSConfiguration class (which implements the CPACS aircraftModelType node). But it should be okay.

Now lets get the other tests working ;)

rainman110 commented:
@RlanderRISCSW Since the 2.2.0 branch is pretty old and was never compiling successfully, I try to fix all problems that occur in the code before your pull request.

rainman110 commented:
@RlanderRISCSW I put OpenCASCADE 6.80 fixes to the 2.2.0 branch. Please sync them with yours (its, best to make a rebase, if possible).

I have no failing tests on my side.

rainman110 commented:
I added some note to my first post regarding CTiglTransformation. Please have a look to it.

RlanderRISCSW commented:
Added support for requesting shapes of CCPACSWingSegment relative to Wing coordinate system, because the CCPACSWingComponentSegment code (and code for internal structures which will be merged separately) works in this coordinate system. The optional CoordinateSystem arguments were added in order to keep the original behavior (return shapes in global coordinates) so that client code does not break.

RlanderRISCSW commented:
I reverted the findSegment method to the original TIGL code (as in release-2.2.0 branch), and added the separate private method findSegmentViaShape. The reason for the second method is that we got problems with the new implementation, which computes the eta/xsi values for the passed point. This computation fails in several cases, e.g.:
> 14:57:27 [ERROR] (CTiglOptimizer.cpp:74) Line search cannot find sufficient decrease, abort
14:57:27 [ERROR] (CTiglPointTranslator.cpp:172) Error in CTiglPointTranslator::translate: p = 11333.930200000001,0,-1212.05 x1 = 10916.5,0,-1212.05 x2 = 10916.5,747,-1212.05 x3 = 14711.32,0,-1212.05 x4 = 14709.330801269476,747,-1334.9049768693042
14:57:27 [ERROR] (CTiglPointTranslator.cpp:179) Resulting eta/xsi: 1.28885e-014,0.11

This is most probably caused by the use of millimeter-units in our CPACS files. Since this is not CPACS conform we plan to change our models for using meters as units, but this means additional conversions and testing in order to ensure that our other models (which are based on the CPACS geometry) still work (e.g. FE model). Therefore we need a solution which also works with millimeters for the moment.

rainman110 commented:
Hej, almost all tests are working. Well done!

RlanderRISCSW commented:
Modifications are finished now from my point of view. Please have a look at the changes and inform us when further modifications of the code should be made

rainman110 commented on src/exports/CTiglExportStep.cpp:
```diff
@@ -386,8 +389,11 @@ void CTiglExportStep::ExportFusedStep(const std::string& filename)
     STEPControl_Controller::Init();
-    Interface_Static::SetCVal("xstep.cascade.unit", "M");
-    Interface_Static::SetCVal("write.step.unit", "M");
+    // CAS_AES write step models in mm (default value)
```
For us, these changes are problematic since we assume, that our model and all the cpacs paramters are defined in meters instead of millimeters. If you insist on millimeters, then you should provide a tigl api function, that allows setting the global unit.

rainman110 commented on src/fuselage/CCPACSFuselageSegment.cpp:
```diff
@@ -309,6 +310,12 @@ PNamedShape CCPACSFuselageSegment::BuildLoft(void)
     generator.Build();
     TopoDS_Shape loftShape = generator.Shape();
 
+    // fixing shape
```
Can you comment, why these shaping fixing is required?

rainman110 commented on src/wing/CCPACSWingComponentSegment.h:
```diff
@@ -51,7 +56,7 @@ typedef std::vector<CCPACSWingSegment*>       SegmentList;
 class CCPACSWing;
 
-class CCPACSWingComponentSegment : public CTiglAbstractSegment
+class CCPACSWingComponentSegment : public CTiglAbstractSegment, public CCPACSWingComponentSegmentLineFunctionality
```
By Subclassing from CCPACSWingComponentSegmentLineFunctionality, you still blow up the API from WingComponentSegment. This does not improve the situation compared to the previous version, it just looks cleaner on the first sight.

Better would be a completely separate Class for CCPACSWingComponentSegmentLineFunctionality, that uses the ComponentSegment as a member or input for its computations.

RlanderRISCSW commented:
Remarks have been checked and according code changes have been reverted. We are finished for now, please check again and tell us in case further code modifications should be executed
